test(linter): print all failing linter test cases

camchenry opened the pull request:
- Closes https://github.com/oxc-project/oxc/issues/16845

Prints a list of all the test cases that failed, instead of just failing on the first one.

- Prints a numbered list and total of cases that failed, for each type of test that failed (expect to pass, expect to fail, expect to fix)
- Prints source code + config for each failure
- Truncates long source code if applicable, and shortens single-line code to be inline

## Example output


```
---- rules::eslint::use_isnan::test stdout ----

22 test cases expected to pass, but failed:

   1. switch(foo) {}
      config: [{"enforceForSwitchCase":true}]
   2. switch(foo) { case bar: NaN; }
      config: [{"enforceForSwitchCase":true}]
   3. switch(foo) { default: NaN; }
      config: [{"enforceForSwitchCase":true}]
   4. switch(Nan) {}
      config: [{"enforceForSwitchCase":true}]
   5. switch('NaN') { default: break; }
      config: [{"enforceForSwitchCase":true}]
   6. switch(foo(NaN)) {}
      config: [{"enforceForSwitchCase":true}]
   7. switch(foo.NaN) {}
      config: [{"enforceForSwitchCase":true}]
   8. switch(foo) { case Nan: break }
      config: [{"enforceForSwitchCase":true}]
   9. switch(foo) { case 'NaN': break }
      config: [{"enforceForSwitchCase":true}]
  10. switch(foo) { case foo(NaN): break }
      config: [{"enforceForSwitchCase":true}]
  11. switch(foo) { case foo.NaN: break }
      config: [{"enforceForSwitchCase":true}]
  12. switch(foo) { case bar: break; case 1: break; default: break; }
      config: [{"enforceForSwitchCase":true}]
  13. switch(foo) { case bar: Number.NaN; }
      config: [{"enforceForSwitchCase":true}]
  14. switch(foo) { default: Number.NaN; }
      config: [{"enforceForSwitchCase":true}]
  15. switch(Number.Nan) {}
      config: [{"enforceForSwitchCase":true}]
  16. switch('Number.NaN') { default: break; }
      config: [{"enforceForSwitchCase":true}]
  17. switch(foo(Number.NaN)) {}
      config: [{"enforceForSwitchCase":true}]
  18. switch(foo.Number.NaN) {}
      config: [{"enforceForSwitchCase":true}]
  19. switch(foo) { case Number.Nan: break }
      config: [{"enforceForSwitchCase":true}]
  20. switch(foo) { case 'Number.NaN': break }
      config: [{"enforceForSwitchCase":true}]
  21. switch(foo) { case foo(Number.NaN): break }
      config: [{"enforceForSwitchCase":true}]
  22. switch(foo) { case foo.Number.NaN: break }
      config: [{"enforceForSwitchCase":true}]



10 test cases expected to fail, but passed:

   1. 123 == NaN;
   2. 123 === NaN;
   3. 123 != NaN;
   4. 123 !== NaN;
   5. 123 == Number.NaN;
   6. 123 === Number.NaN;
   7. 123 != Number.NaN;
   8. 123 !== Number.NaN;
   9. x === Number?.NaN;
  10. x === Number['NaN'];



8 fixes did not produce expected output:

   1.    Input: 1 == NaN
      Expected: isNaN(1)
        Actual: <test passed, no fix applied>

   2.    Input: 1 === NaN
      Expected: isNaN(1)
        Actual: <test passed, no fix applied>

   3.    Input: 1 != NaN
      Expected: !isNaN(1)
        Actual: <test passed, no fix applied>

   4.    Input: 1 !== NaN
      Expected: !isNaN(1)
        Actual: <test passed, no fix applied>

   5.    Input: 1 == Number.NaN
      Expected: isNaN(1)
        Actual: <test passed, no fix applied>

   6.    Input: 1 === Number.NaN
      Expected: isNaN(1)
        Actual: <test passed, no fix applied>

   7.    Input: 1 != Number.NaN
      Expected: !isNaN(1)
        Actual: <test passed, no fix applied>

   8.    Input: 1 !== Number.NaN
      Expected: !isNaN(1)
        Actual: <test passed, no fix applied>
```

```
---- rules::typescript::prefer_literal_enum_member::test stdout ----

thread 'rules::typescript::prefer_literal_enum_member::test' (883345) panicked at crates/oxc_linter/src/tester.rs:414:13:

8 test cases expected to pass but failed:

   1. | 
      │                 enum ValidRegex {
      │                 A = /test/,
      │             }
      │                     

   2. | 
      │                 enum ValidString {
      │                   A = 'test',
      │                 }
      │                     

   3. | 
      │                 enum ValidNumber {
      │                   A = 42,
      │                 }
      │                     

   4. | 
      │                 enum ValidNumber {
      │                   A = -42,
      │                 }
      │                     

   5. | 
      │                 enum ValidNumber {
      │                   A = +42,
      │                 }
      │                     

   6. | 
      │                 enum ValidNull {
      │                   A = null,
      │                 }
      │                     

   7. | 
      │                 enum ValidQuotedKeyWithAssignment {
      │                   'a' = 1,
      │                 }
      │                     

   8. | 
      │                 enum Foo {
      │                   A = 1 << 0,
      │                   B = 1 >> 0,
      │                   C = 1 >>> 0,
      │                   D = 1 | 0,
      │                   E = 1 & 0,
      │                   F = 1 ^ 0,
      │                   G = ~1,
      │                 }
      │ ... (1 more line(s))

      config: [{"allowBitwiseExpressions":true}]
```

```
---- rules::eslint::yoda::test stdout ----

thread 'rules::eslint::yoda::test' (883045) panicked at crates/oxc_linter/src/tester.rs:430:13:

16 test cases expected to fail but passed:

   1. if (x <= 'foo' || 'bar' < x) {}
      config: ["always",{"exceptRange":true}]
   2. if ("red" == value) {}
      config: ["never"]
   3. if (true === value) {}
      config: ["never"]
   4. if (null !== value) {}
      config: ["never"]
   5. if ("red" <= value) {}
      config: ["never"]
   6. if (true >= value) {}
      config: ["never"]
   7. function foo() { return (null > value); }
      config: ["never"]
   8. if (value == "red") {}
      config: ["always"]
   9. if (value === true) {}
      config: ["always"]
  10. x < ('foo')in bar
      config: ["always"]
  11. false <= ((x))in foo
      config: ["never"]
  12. false <= ((x)) in foo
      config: ["never"]
  13. if('a' <= x && x < 'b') {}
      config: ["always"]
  14. if ('b' <= x && x < 'a') {}
      config: ["never",{"exceptRange":true}]
  15. if('a' <= x && x < 1) {}
      config: ["never",{"exceptRange":true}]
  16. {( t=='' )}
      config: ["always",{"onlyEquality":true}]
```

```
---- rules::eslint::no_unneeded_ternary::test stdout ----

thread 'rules::eslint::no_unneeded_ternary::test' (939816) panicked at crates/oxc_linter/src/tester.rs:537:13:

3 fixes did not produce expected output:

   1.    Input: var a = foo ? foo : 'No';
      Expected: var a = foo || 'No';
        Actual: var a = foo || ('No');

   2.    Input: f(x ? x : 1);
      Expected: f(x || 1);
        Actual: f(x || (1));

   3.    Input: x ? x : 1;
      Expected: x || 1;
        Actual: x || (1);
```

camchenry commented:
Hm, those snapshots seem off actually 🤔

camc314 commented on crates/oxc_linter/src/tester.rs:
```diff
@@ -369,10 +370,34 @@ impl Tester {
+        let fix_failures = self.test_fix();
+
+        if let Err(failed) = &failed
+            && !failed.is_empty()
```
it's probably overkill for this to return `Result<(), Vec>`

on L385 when you check `passed.is_empty()`, this is always true if `passed` is the `Err` case - we can either drop ths `!passed.is_empty()` check or just return a `Result` rather than `Result<vec>`

camc314 reviewed (approved):
💪 👍 nice!

camchenry committed:
test(linter): print all failing linter test cases (#16937)

- Closes https://github.com/oxc-project/oxc/issues/16845

Prints a list of all the test cases that failed, instead of just failing on the first one.

- Prints a numbered list and total of cases that failed, for each type of test that failed (expect to pass, expect to fail, expect to fix)
- Prints source code + config for each failure
- Truncates long source code if applicable, and shortens single-line code to be inline

## Example output

```
---- rules::eslint::use_isnan::test stdout ----

22 test cases expected to pass, but failed:

   1. switch(foo) {}
      config: [{"enforceForSwitchCase":true}]
   2. switch(foo) { case bar: NaN; }
      config: [{"enforceForSwitchCase":true}]
   3. switch(foo) { default: NaN; }
      config: [{"enforceForSwitchCase":true}]
   4. switch(Nan) {}
      config: [{"enforceForSwitchCase":true}]
   5. switch('NaN') { default: break; }
      config: [{"enforceForSwitchCase":true}]
   6. switch(foo(NaN)) {}
      config: [{"enforceForSwitchCase":true}]
   7. switch(foo.NaN) {}
      config: [{"enforceForSwitchCase":true}]
   8. switch(foo) { case Nan: break }
      config: [{"enforceForSwitchCase":true}]
   9. switch(foo) { case 'NaN': break }
      config: [{"enforceForSwitchCase":true}]
  10. switch(foo) { case foo(NaN): break }
      config: [{"enforceForSwitchCase":true}]
  11. switch(foo) { case foo.NaN: break }
      config: [{"enforceForSwitchCase":true}]
  12. switch(foo) { case bar: break; case 1: break; default: break; }
      config: [{"enforceForSwitchCase":true}]
  13. switch(foo) { case bar: Number.NaN; }
      config: [{"enforceForSwitchCase":true}]
  14. switch(foo) { default: Number.NaN; }
      config: [{"enforceForSwitchCase":true}]
  15. switch(Number.Nan) {}
      config: [{"enforceForSwitchCase":true}]
  16. switch('Number.NaN') { default: break; }
      config: [{"enforceForSwitchCase":true}]
  17. switch(foo(Number.NaN)) {}
      config: [{"enforceForSwitchCase":true}]
  18. switch(foo.Number.NaN) {}
      config: [{"enforceForSwitchCase":true}]
  19. switch(foo) { case Number.Nan: break }
      config: [{"enforceForSwitchCase":true}]
  20. switch(foo) { case 'Number.NaN': break }
      config: [{"enforceForSwitchCase":true}]
  21. switch(foo) { case foo(Number.NaN): break }
      config: [{"enforceForSwitchCase":true}]
  22. switch(foo) { case foo.Number.NaN: break }
      config: [{"enforceForSwitchCase":true}]

10 test cases expected to fail, but passed:

   1. 123 == NaN;
   2. 123 === NaN;
   3. 123 != NaN;
   4. 123 !== NaN;
   5. 123 == Number.NaN;
   6. 123 === Number.NaN;
   7. 123 != Number.NaN;
   8. 123 !== Number.NaN;
   9. x === Number?.NaN;
  10. x === Number['NaN'];

8 fixes did not produce expected output:

   1.    Input: 1 == NaN
      Expected: isNaN(1)
        Actual: <test passed, no fix applied>

   2.    Input: 1 === NaN
      Expected: isNaN(1)
        Actual: <test passed, no fix applied>

   3.    Input: 1 != NaN
      Expected: !isNaN(1)
        Actual: <test passed, no fix applied>

   4.    Input: 1 !== NaN
      Expected: !isNaN(1)
        Actual: <test passed, no fix applied>

   5.    Input: 1 == Number.NaN
      Expected: isNaN(1)
        Actual: <test passed, no fix applied>

   6.    Input: 1 === Number.NaN
      Expected: isNaN(1)
        Actual: <test passed, no fix applied>

   7.    Input: 1 != Number.NaN
      Expected: !isNaN(1)
        Actual: <test passed, no fix applied>

   8.    Input: 1 !== Number.NaN
      Expected: !isNaN(1)
        Actual: <test passed, no fix applied>
```

```
---- rules::typescript::prefer_literal_enum_member::test stdout ----

thread 'rules::typescript::prefer_literal_enum_member::test' (883345) panicked at crates/oxc_linter/src/tester.rs:414:13:

8 test cases expected to pass but failed:

   1. |
      │                 enum ValidRegex {
      │                 A = /test/,
      │             }
      │

   2. |
      │                 enum ValidString {
      │                   A = 'test',
      │                 }
      │

   3. |
      │                 enum ValidNumber {
      │                   A = 42,
      │                 }
      │

   4. |
      │                 enum ValidNumber {
      │                   A = -42,
      │                 }
      │

   5. |
      │                 enum ValidNumber {
      │                   A = +42,
      │                 }
      │

   6. |
      │                 enum ValidNull {
      │                   A = null,
      │                 }
      │

   7. |
      │                 enum ValidQuotedKeyWithAssignment {
      │                   'a' = 1,
      │                 }
      │

   8. |
      │                 enum Foo {
      │                   A = 1 << 0,
      │                   B = 1 >> 0,
      │                   C = 1 >>> 0,
      │                   D = 1 | 0,
      │                   E = 1 & 0,
      │                   F = 1 ^ 0,
      │                   G = ~1,
      │                 }
      │ ... (1 more line(s))

      config: [{"allowBitwiseExpressions":true}]
```

```
---- rules::eslint::yoda::test stdout ----

thread 'rules::eslint::yoda::test' (883045) panicked at crates/oxc_linter/src/tester.rs:430:13:

16 test cases expected to fail but passed:

   1. if (x <= 'foo' || 'bar' < x) {}
      config: ["always",{"exceptRange":true}]
   2. if ("red" == value) {}
      config: ["never"]
   3. if (true === value) {}
      config: ["never"]
   4. if (null !== value) {}
      config: ["never"]
   5. if ("red" <= value) {}
      config: ["never"]
   6. if (true >= value) {}
      config: ["never"]
   7. function foo() { return (null > value); }
      config: ["never"]
   8. if (value == "red") {}
      config: ["always"]
   9. if (value === true) {}
      config: ["always"]
  10. x < ('foo')in bar
      config: ["always"]
  11. false <= ((x))in foo
      config: ["never"]
  12. false <= ((x)) in foo
      config: ["never"]
  13. if('a' <= x && x < 'b') {}
      config: ["always"]
  14. if ('b' <= x && x < 'a') {}
      config: ["never",{"exceptRange":true}]
  15. if('a' <= x && x < 1) {}
      config: ["never",{"exceptRange":true}]
  16. {( t=='' )}
      config: ["always",{"onlyEquality":true}]
```

```
---- rules::eslint::no_unneeded_ternary::test stdout ----

thread 'rules::eslint::no_unneeded_ternary::test' (939816) panicked at crates/oxc_linter/src/tester.rs:537:13:

3 fixes did not produce expected output:

   1.    Input: var a = foo ? foo : 'No';
      Expected: var a = foo || 'No';
        Actual: var a = foo || ('No');

   2.    Input: f(x ? x : 1);
      Expected: f(x || 1);
        Actual: f(x || (1));

   3.    Input: x ? x : 1;
      Expected: x || 1;
        Actual: x || (1);
```
